tools/import-verifier/import-verifier.go

```diff
@@ -0,0 +1,164 @@
+	forbiddenImportSet := map[string]struct{}{}
+	for _, imp := range append(pkg.Imports, append(pkg.TestImports, pkg.XTestImports...)...) {
+		if i.isForbidden(imp) {
+			forbiddenImportSet[relativePackage(imp)] = struct{}{}
+		}
+	}
+
+	var forbiddenImports []string
+	for imp := range forbiddenImportSet {
+		forbiddenImports = append(forbiddenImports, imp)
+	}
+	return forbiddenImports
+}
+
+// isForbidden determines if an import is forbidden,
+// which is true when the import is:
+//   - of a package under the rootPackage
+//   - is not of the base import path or a sub-package of it
+//   - is not of an allowed path or a sub-package of one
+func (i *ImportRestriction) isForbidden(imp string) bool {
+	importsBelowRoot := strings.HasPrefix(imp, rootPackage)
+	importsBelowBase := strings.HasPrefix(imp, absolutePackage(i.BaseImportPath))
+	importsAllowed := false
+	for _, allowed := range i.AllowedImports {
+		importsAllowed = importsAllowed || strings.HasPrefix(imp, absolutePackage(allowed))
+	}
+
+	return importsBelowRoot && !importsBelowBase && !importsAllowed
+}
+
+func absolutePackage(relativePackage string) string {
+	return fmt.Sprintf("%s/%s", rootPackage, relativePackage)
+}
+
+func relativePackage(absolutePackage string) string {
+	return absolutePackage[len(rootPackage)+1:]
+}
+
+func main() {
+	if len(os.Args) != 2 {
+		log.Fatalf("%s requires the configuration file as it's only argument", os.Args[0])
+	}
+
+	configFile := os.Args[1]
+	importRestrictions, err := loadImportRestrictions(configFile)
+	if err != nil {
+		log.Fatalf("Failed to load import restrictions: %v", err)
+	}
+
+	foundForbiddenImports := false
+	for _, restriction := range importRestrictions {
+		log.Printf("Inspecting imports under %s...\n", restriction.BaseImportPath)
+		packages, err := resolvePackageTree(absolutePackage(restriction.BaseImportPath))
+		if err != nil {
+			log.Fatalf("Failed to resolve package tree: %v", err)
+		}
+
+		log.Printf("-- validating imports for %d packages in the tree", len(packages))
+		for _, pkg := range packages {
+			if forbidden := restriction.ForbiddenImportsFor(pkg); len(forbidden) != 0 {
+				logForbiddenPackages(relativePackage(pkg.ImportPath), forbidden)
+				foundForbiddenImports = true
+			}
+		}
+	}
+
+	if foundForbiddenImports {
+		os.Exit(1)
+	}
+}
+
+func loadImportRestrictions(configFile string) ([]ImportRestriction, error) {
+	config, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load configuration from %s: %v", configFile, err)
+	}
+
+	var importRestrictions []ImportRestriction
+	if err := json.Unmarshal(config, &importRestrictions); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal from %s: %v", configFile, err)
+	}
+
+	return importRestrictions, nil
+}
+
+func resolvePackageTree(treeBase string) ([]Package, error) {
+	cmd := "go"
+	args := []string{"list", "-json", fmt.Sprintf("%s...", treeBase)}
+	stdout, err := exec.Command(cmd, args...).Output()
```

Comment:
Yeah, real bummer that we don't get a Go API for this, but grabbing the JSON and decoding it is good enough.